### PR TITLE
Keyboard events on search

### PR DESCRIPTION
**Pull request by MarcoPolo**
@keybase/react-hackers

### CODE REVIEW

**Comment by keybase-ci-visdiff** [comment]
The commits 86fb03cf9501...0aecb305a214 introduce visual changes on linux. <details><summary>:mag_right: 1 changed</summary>

 * Changed: **Tracker-New_user_Scroll_1** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1497047064112-86fb03cf9501-0aecb305a214-linux/Tracker-New_user_Scroll_1-86fb03cf9501.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1497047064112-86fb03cf9501-0aecb305a214-linux/Tracker-New_user_Scroll_1-0aecb305a214.png)  
   Tracker-New_user_Scroll_1 comparison
</summary>

File: shared/actions/chat/index.js [hunk 1]
```diff
@@ -1016,6 +1016,33 @@ function* _exitSearch() {
...
+  yield put(Creators.clearSearchResults())
+}
+
+function* _selectFirstSearchResult() {
+  const searchResults = yield select(({chat: {searchResults}}: TypedState) => searchResults)
+  yield put(Creators.selectSearchResultId(searchResults.first()))
+}
+
+function* _clearSelectedSearchResult() {
+  yield put(Creators.selectSearchResultId(null))
```
Comment (chrisnojima) [inline shared/actions/chat/index.js]: don't these need some kind of id or key path to know which search you'd mutating? like won't this affect profile and chat at the same time or something

Comment (MarcoPolo) [inline shared/actions/chat/index.js]: no this lives solely in the chat store. This isn't an entity, it's like the list of what search result ids to show. It's some state that affects the ui.

File: shared/actions/chat/index.js [hunk 2]
```diff
@@ -1016,6 +1016,33 @@ function* _exitSearch() {
   yield put(Creators.setInboxFilter([]))
 }
 
+function* _selectSearchResultWithMovement({payload: {movement}}: Constants.SelectSearchResultIdWithMovement) {
+  const searchResults = yield select(({chat: {searchResults}}: TypedState) => searchResults)
+  const selectedSearchId = yield select(({chat: {selectedSearchId}}: TypedState) => selectedSearchId)
+  const index = searchResults.indexOf(selectedSearchId)
```
Comment (chromakode) [inline shared/actions/chat/index.js]: What if for some reason this returns `-1`? Should there be a warning / reset logic?

File: shared/actions/chat/index.js [hunk 3]
```diff
@@ -1016,6 +1016,33 @@ function* _exitSearch() {
   yield put(Creators.setInboxFilter([]))
 }
 
+function* _selectSearchResultWithMovement({payload: {movement}}: Constants.SelectSearchResultIdWithMovement) {
+  const searchResults = yield select(({chat: {searchResults}}: TypedState) => searchResults)
```
Comment (chromakode) [inline shared/actions/chat/index.js]: [Discussion, can defer from review]: I think that this sort of operation should be done in the reducer, if possible. It's possible for something to run in between these two selectors, causing them to be out of sync. Might be better to combine them into a single selector, but still feels like doing them outside of the saga asyncness is safer.

File: shared/searchv3/results-list/index.desktop.js [hunk 4]
```diff
@@ -23,6 +23,7 @@ class SearchResultsList extends Component<void, Props, void> {
         key={id}
         onClick={() => onClick(id)}
         onShowTracker={onShowTracker ? () => onShowTracker(id) : undefined}
+        selectedId={this.props.selectedId}
```
Comment (chromakode) [inline shared/searchv3/results-list/index.desktop.js]: I think you should do the `id === this.props.selectedId` here and pass in the selected state. This way all rows do not have their props change when the `selectedId` changes, and the rows are not concerned with what the selected id of the collection is.

File: shared/actions/chat/index.js (hunk 1 above)
Comment (chrisnojima) [inline shared/actions/chat/index.js]: but wont we need this kind of logic for this elsewhere. 
it seems like the up/down changing the selection is generic shared code (with keypath) and what happens as a side effect of the selection changing is chat specific

**Comment by keybase-ci-visdiff** [comment]
:no_entry: Error rendering the commits 69f9785e56e7...45d6b3d5dff1 on linux. <details><summary>:mag_right: 6 errored</summary>

 * Error in 69f9785e56e7: **Signup_Passphrase-Start-ERROR**  
   Signup_Passphrase-Start-ERROR rendered
 * Error in 45d6b3d5dff1: **Signup_Passphrase-Start-ERROR**  
   Signup_Passphrase-Start-ERROR rendered
 * Error in 69f9785e56e7: **Signup_Passphrase-Error-ERROR**  
   Signup_Passphrase-Error-ERROR rendered
 * Error in 45d6b3d5dff1: **Signup_Passphrase-Error-ERROR**  
   Signup_Passphrase-Error-ERROR rendered
 * Error in 69f9785e56e7: **Signup_Error-Start-ERROR**  
   Signup_Error-Start-ERROR rendered
 * Error in 45d6b3d5dff1: **Signup_Error-Start-ERROR**  
   Signup_Error-Start-ERROR rendered
</summary>

File: shared/searchv3/helpers.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+import {compose, withHandlers, withPropsOnChange, withState, lifecycle} from 'recompose'
+import * as Constants from '../constants/searchv3'
+import {debounce} from 'lodash'
+
+type OwnProps = {
+  onChangeSearchText: (s: string) => void,
+  search: (term: string, service: Constants.Service) => void,
+  selectedService: Constants.Service,
+
```
Comment (chromakode) [inline shared/searchv3/helpers.js]: :scissors: blank line?

**Pull request merged**